backend/app/notify/email.go

```diff
@@ -0,0 +1,162 @@
+		return &res, errors.Errorf(
+			"[WARN] error establishing test connecting to '%s':%d with username '%s': %s",
+			res.server, res.port, res.username, err)
+	}
+	if err = tmpConn.Close(); err != nil {
+		return &res, errors.Errorf(
+			"[WARN] error closing test connection to %s:%d: %s",
+			res.server, res.port, err)
+	}
+
+	// Activate server goroutine.
+	go res.activate()
+	// Close the channel to stop the mail daemon
+	go func() { <-ctx.Done(); close(res.sendChan) }()
+
+	return &res, nil
+}
+
+// Send email from request to address in settings
+func (e *Email) Send(ctx context.Context, req request) error {
+	log.Printf("[DEBUG] send notification via %s, comment id %s", e, req.comment.ID)
+
+	messageBody := prepareBody(req)
+
+	// Create message.
+	m := gomail.NewMessage()
+	// TODO: figure out where "from" and "to" addresses come from
+	//m.SetHeader("From", "no-reply@example.com")
+	//m.SetAddressHeader("To", req.Address, req.Name)
+	m.SetHeader("Subject", fmt.Sprintf("New comment for \"%s\"", req.comment.PostTitle))
+	m.SetBody("text/html", messageBody)
+
+	// Wait for ability to send message and return error from error channel after sending it.
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case e.sendChan <- m:
+		return <-e.errChan
+	}
+}
+
+func (e *Email) String() string {
+	return fmt.Sprintf("email: '%s'@'%s':%d", e.username, e.server, e.port)
+}
+
+//prepareBody generates email message text based on request
+func prepareBody(req request) string {
+	from := req.comment.User.Name
+	if req.comment.ParentID != "" {
+		from += " → " + req.parent.User.Name
+	}
+	link := fmt.Sprintf("↦ [original comment](%s)", req.comment.Locator.URL+uiNav+req.comment.ID)
+	if req.comment.PostTitle != "" {
+		link = fmt.Sprintf("↦ [%s](%s)", req.comment.PostTitle, req.comment.Locator.URL+uiNav+req.comment.ID)
+	}
+	body := fmt.Sprintf("%s\n\n%s\n\n%s", from, req.comment.Orig, link)
+	return body
+}
+
+func (e *Email) activate() {
+	d := gomail.NewDialer(e.server, e.port, e.username, e.password)
+
+	var s gomail.SendCloser
+	var err error
+	open := false
+	for {
+		select {
+		case m, ok := <-e.sendChan:
+			if !ok {
+				close(e.errChan)
+				return
+			}
+			if !open {
+				if s, err = d.Dial(); err != nil {
+					// Problems with connection, returning error and considering connection not established.
+					e.errChan <- errors.Errorf(
+						"error connecting to %s:%d with username %s: %s",
+						e.server, e.port, e.username, err)
+					break
+				}
+				open = true
+			}
+			if err = gomail.Send(s, m); err != nil {
+				e.errChan <- errors.Errorf(
+					"error sending to %s:%d: %s",
+					e.server, e.port, err)
+			}
+			// Email sent successfully.
+			e.errChan <- nil
+		// Close the connection to the SMTP server if no email was sent in the keepAlive period.
+		case <-time.After(e.keepAlive):
+			if open {
+				if err := s.Close(); err != nil {
+					// Problems with closing connection, considering connection still established.
```

Comment:
I'm not sure this is a good idea. If we failed to close connection we can't consider that connection well and alive

Reply:
Currently, it's somehow similar to `panic` (as it was originally): we just shut down the server in case of connection problems. That's why it have `break` inside.